Gtk3

01micko opened the pull request:
I'm attempting to use the correct options to `gtkdialog` to call a `css` file instead of my nasty hack of resetting `XDG_CONFIG_HOME`

Still a work in progress.

Commit:
remove gtk3_theme_otf

use a new mechanism for on the fly gtk3 themeing in gtkdialog

Commit:
improve /usr/lib/gtkdialog/box_splash re: #2248

Commit:
restore thr ptheme sourcing

Commit:
update ptheme re #2248

Commit:
update pup-advert-blocker re #2248

Commit:
last 2 commits depend on updated xml_info

also ignore xml_scalegrip if gtkdialog uses gtk3

Commit:
update quicksetup re #2248

Commit:
fix breakage of last commit

Commit:
update first batch in /usr/sbin re #2248

Commit:
more fixes in /usr/sbin re #2248

Commit:
workaround for scalegrip in gtk3

Commit:
fix the rest of scripts in skeleton re #2248

Commit:
update some of the jwm_config utils re: #2248

peabee commented:
Sounds like a good idea as long as it doesn't break Gtk2 builds...... Not sure what we can do to help other than test builds once they appear .......??

zigbert commented:
I am in lack of time to take a deep dive these days, but my brief look into the code looks very promising. I like the --styles solution which is clean and compatible with gtk2. There is still some issues to solve, but you have set the framework, which is the key for success.

dimkr commented:
Triggered https://github.com/puppylinux-woof-CE/woof-CE/actions/runs/801826375, so we have GTK 2 (Slacko 7.x) and GTK 3 (Slacko 8.x) builds for bug hunting and comparison.

peabee commented:
> Triggered https://github.com/puppylinux-woof-CE/woof-CE/actions/runs/801826375, so we have GTK 2 (Slacko 7.x) and GTK 3 (Slacko 8.x) builds for bug hunting and comparison.

Gtk3 aspects of Slacko64-8.1 seem much better although I feel quicksetup window is rather too wide....

@01micko - I notice that Slacko64-8.1 does not recognise Arch64 pkg.tar.zst packages....

![Screenshot](https://user-images.githubusercontent.com/5926865/116788723-27db3280-aaa3-11eb-810c-8a981afaf4cf.png)

peabee commented:
p.s. why do the iso's and puppy.sfs of seemingly identical builds vary in size so much?

dimkr commented:
> p.s. why do the iso's and puppy.sfs of seemingly identical builds vary in size so much?

You're probably talking about https://github.com/puppylinux-woof-CE/woof-CE/blob/testing/.github/workflows/build.yml#L79

Commit:
Merge branch 'testing' into gtk3

01micko commented:
> I notice that Slacko64-8.1 does not recognise Arch64 pkg.tar.zst packages

@peabee are you using Slackware's `shared-mime-info`?

Commit:
improve pupdialog re #2248

01micko commented on woof-code/rootfs-skeleton/usr/bin/pupdialog:
```diff
@@ -208,11 +208,27 @@ done
+}
+button.radio:checked {
+	border: 3px solid ${FOREGROUND};
+}
```
```
pupdialog --backtitle "backtitle here" --background yellow --foreground grey20  --default-item tag2  --menu "this is blurb for menu dlg" 0 0 0 tag1 item1 tag2 item2 tag3 item3
```

![pupdialog](https://user-images.githubusercontent.com/1019119/116796837-30496280-aacf-11eb-8e88-e4788b1b4eea.png)

Commit:
separate back out the wallpaper setter from rox_config

Commit:
Update wallpaper.desktop

Commit:
adjust pclock re #2248

I believe it also looks better now in gtk2. I've hidden the top vbox whch was necessary for gtk3 but works well in gtk2.

Commit:
update ptiming re #2248

very similar to pclock and share much code

01micko commented:
next stop .. `pmusic`

It might take a while.

peabee commented:
> @peabee are you using Slackware's `shared-mime-info`?

No - my builds use common32 / 64 - version 1.10

Slacko64-8.1 is using very old version 1.2:
shared-mime-info-1.2-x86_64_s700|shared-mime-info|1.2-x86_64_s700||BuildingBlock|4976||shared-mime-info-1.2-x86_64_s700.pet||mimetype information|slackware64|14.2||
shared-mime-info_DEV-1.2-x86_64_s700|shared-mime-info_DEV|1.2-x86_64_s700||BuildingBlock|32||shared-mime-info_DEV-1.2-x86_64_s700.pet|+shared-mime-info|shared-mime-info development||||

zigbert commented:
> next stop .. `pmusic`
> 
> It might take a while.

Looking forward to it :)

01micko commented:
> > @peabee are you using Slackware's `shared-mime-info`?
> 
> No - my builds use common32 / 64 - version 1.10
> 
> Slacko64-8.1 is using very old version 1.2:
> shared-mime-info-1.2-x86_64_s700|shared-mime-info|1.2-x86_64_s700||BuildingBlock|4976||shared-mime-info-1.2-x86_64_s700.pet||mimetype information|slackware64|14.2||
> shared-mime-info_DEV-1.2-x86_64_s700|shared-mime-info_DEV|1.2-x86_64_s700||BuildingBlock|32||shared-mime-info_DEV-1.2-x86_64_s700.pet|+shared-mime-info|shared-mime-info development||||

Thanks Peter, I'll correct that shortly.

Commit:
Merge branch 'testing' into gtk3

01micko commented:
@peabee `shared-mime-info-1.10` is still too old. I had success with slackware's 2.1 but there is still work to do.

As this is off topic I created issue #2251

Commit:
Merge branch 'testing' into gtk3

Commit:
Merge branch 'testing' into gtk3

Commit:
Merge branch 'testing' into gtk3

Commit:
remove debug from time script

Commit:
Merge branch 'testing' into gtk3

Commit:
Merge branch 'testing' into gtk3

01micko commented:
@zigbert ziggy stardust is back

@dimkr 

![widgets1](https://user-images.githubusercontent.com/1019119/117552521-60e34c00-b03b-11eb-8b53-7a8e1d08a2b1.png)

![widgets2](https://user-images.githubusercontent.com/1019119/117552524-65a80000-b03b-11eb-9c4b-8a5bb2893f98.png)

![widgets3](https://user-images.githubusercontent.com/1019119/117552526-6a6cb400-b03b-11eb-9bd8-cb716872a1e7.png)

zigbert commented:
Mmmm - you really get the hang of it ;) - Some stated that Stardust was 'impossible' in GTK3

I try to crack some bugs in my small drops of free time...
We are opening a new gallery in a near future. Not the same artwork that you are working on, but still...

01micko commented:
The problem is the png files as in gtk3 they don't stretch so you have to 'draw' images in pure CSS. The `stardust_bright` gtk3 theme is certainly _not_ a faithful reproduction of the gtk2 version but it's getting closer.
Actually, I opened this browser before I fixed the theme and you can see what I mean with the png files in the scrollbar.

![scroll-bug](https://user-images.githubusercontent.com/1019119/117554486-9857f580-b047-11eb-9048-83a10acf49a7.png)


I'm going to commit it today and hopefully the 'touch' version as well then we go for the `stardust_dark`.

01micko commented:
@zigbert 

> We are opening a new gallery in a near future. Not the same artwork that you are working on, but still...

That's great news! Good Luck :four_leaf_clover:

01micko commented:
@dimkr - just thought I'd mention that commit I made to gtkdialog - [3c18c9e](https://github.com/puppylinux-woof-CE/gtkdialog/commit/3c18c9e0f1d00a30531fef16058b73f3a4fa49cf)

The maximum chars can easily be overidden with `width-chars`:
```xml
 <text width-chars="120">
```

01micko commented:
@dimkr - just thought I'd mention that commit I made to gtkdialog - [3c18c9e](https://github.com/puppylinux-woof-CE/gtkdialog/commit/3c18c9e0f1d00a30531fef16058b73f3a4fa49cf)

The maximimum chars can easily be overidden with `width-chars`:
```xml
 <text width-chars="120">
```

01micko commented:
@zigbert 

Since there are complexities in the theming of PMusic it will take me some time to get up to speed with gtk3. Probably several months. 

Also, pequalizer is giving problems - mostly due to the addition of pulseaudio. No big deal - it had a good run. Maybe I can make one based on pulse - but that's for another time. The upshot of that is that I won't be including pequalizer in future versions if pulseaudio is installed. 

PMusic runs for me just fine if I disable pequalizer and drop in a gtkdialog based on gtk2 - no problem if I build it without VTE support. I just renamed the exec to `gtk2dialog` so as not to conflict.

My request to you is to add this small patch to PMusic (and friends) so that I can use the alternate exec. The patch will be temporary until we get gtk3 theming in PMusic stable.

_PATCH REMOVED_ - see post below

TIA

**EDIT**
On further inspection I see `gtkdialog` is hard coded in more places than the patch mentions, especially in `ps` calls. Do they work with the variable `$GTKDIALOG`?

Commit:
Merge branch 'testing' into gtk3

Commit:
bugfix startdust_bright gtk2 theme

Commit:
add startdust_bright_mouse gtk3 theme

Commit:
add stardust_bright_touch gtk3 theme re #2248

01micko commented:
@zigbert 

Here is a revised patch, seems to work ok.. no code is touched just hard coding replaced with variable - and `GTK3` test added in another spot.

```diff
diff -rupN a/usr/local/pmusic/box_input b/usr/local/pmusic/box_input
--- a/usr/local/pmusic/box_input	2020-07-09 08:32:28.000000000 +1000
+++ b/usr/local/pmusic/box_input	2021-05-09 11:37:10.763115364 +1000
@@ -2,7 +2,13 @@
 $APPDIR/func -reset_gtk
 export TEXTDOMAIN=pmusic
 export OUTPUT_CHARSET=UTF-8
-[ ! "$GTKDIALOG" ] && GTKDIALOG="gtkdialog" #puppy specific
+if [ ! "$GTKDIALOG" ];then #puppy specific
+	if [ "$GTKDIALOG_BUILD" = 'GTK3' ];then
+		export GTKDIALOG=gtk2dialog
+	else
+		export GTKDIALOG=gtkdialog
+	fi
+fi
 if [ -z "$FRAME" ]; then FRAME=" "; fi
 export WORKDIR=$HOME/.pmusic/tmp
 
diff -rupN a/usr/local/pmusic/func b/usr/local/pmusic/func
--- a/usr/local/pmusic/func	2021-03-03 12:38:35.000000000 +1000
+++ b/usr/local/pmusic/func	2021-05-09 11:24:33.253129431 +1000
@@ -141,7 +141,7 @@ pMusic will work just fine even if missi
 	   </tree>
 	  </vbox>
 	 </vbox>
-	</window>' | gtkdialog -s &
+	</window>' | $GTKDIALOG -s &
 	;;
 
 -visualization)
@@ -773,7 +773,7 @@ $(gettext 'Requested filter:') $TMP"
 	#first quit guis
 	TMP="$(ps -eo pid,command)"
 	PS=$(grep -iE "pmusic|aplay|ffplay" <<< "$TMP")
-	for I in `grep -F gtkdialog <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done #do not kill radiohits grabber 
+	for I in `grep -F $GTKDIALOG <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done #do not kill radiohits grabber 
 	#stop audio
 	pmusic -s stop
 	rm -f $WORKDIR/gui_is_running #ensure to flag 'no gui'
@@ -797,7 +797,7 @@ $(gettext 'Requested filter:') $TMP"
 	for I in `grep -F index <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -F aplay <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -F playing <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
-	for I in `grep -F gtkdialog <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
+	for I in `grep -F $GTKDIALOG <<< "$PS" | grep -vF radiohits | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	for I in `grep -vF radiohits <<< "$PS" | awk '{print $1}' 2>> $STDERR`; do kill -9 $I 2>> $STDERR; done
 	killall cdda2wav 2>> $STDERR
 	#reset info about search filters. Next run will use settings from preferences
diff -rupN a/usr/local/pmusic/gui_trackinfo b/usr/local/pmusic/gui_trackinfo
--- a/usr/local/pmusic/gui_trackinfo	2020-07-09 08:32:28.000000000 +1000
+++ b/usr/local/pmusic/gui_trackinfo	2021-05-09 11:30:14.783123061 +1000
@@ -90,7 +90,7 @@ S='
   <button tooltip-text="'$(gettext 'Search for alternative meta information based on ARTIST and TITLE. The search-string does not require complete Artist/Title information, but you MUST use complete words.')'">
    <input file stock="gtk-find"></input>
    <action>'$APPDIR'/func_trackinfo "'${MYPID}'" -search_alternatives</action>
-   <action>gtkdialog -f '$WORKDIR'/xml_pmusic_trackinfo'${MYPID}'_alternativ_id3 &</action>
+   <action>'$GTKDIALOG' -f '$WORKDIR'/xml_pmusic_trackinfo'${MYPID}'_alternativ_id3 &</action>
   </button>
  </hbox>
  
diff -rupN a/usr/local/pmusic/pmusic b/usr/local/pmusic/pmusic
--- a/usr/local/pmusic/pmusic	2021-03-03 12:38:35.000000000 +1000
+++ b/usr/local/pmusic/pmusic	2021-05-09 12:04:41.803084723 +1000
@@ -23,7 +23,11 @@ export OUTPUT_CHARSET=UTF-8
 #---
 export AUDIO_FORMATS="\.3gp$|\.aac$|\.ac3$|\.act$|\.aif$|\.aiff$|\.amr$|\.ape$|\.asf$|\.au$|\.dss$|\.flac$|\.gsm$|\.kar$|\.m4a$|\.m4b$|\.mid$|\.mmf$|\.mp1$|\.mp2$|\.mp3$|\.mp4$|\.mpc$|\.oga$|\.ogg$|\.opus$|\.ra$|\.raw$|\.rm$|\.shn$|\.sln$|\.tta$|\.voc$|\.wav$|\.webm$|\.wma$|\.wv$"
 type pequalizer > /dev/null 2>&1 && pequalizer -s & #setup initial presets
-export GTKDIALOG=gtkdialog
+if [ "$GTKDIALOG_BUILD" = 'GTK3' ];then
+	export GTKDIALOG=gtk2dialog
+else
+	export GTKDIALOG=gtkdialog
+fi
 export TXT_SEARCH="$(gettext 'Searching more info')"
 export TXT_NO_WWW="$(gettext 'Not able to connect to internet')"
 export SF='space-expand="false" space-fill="false"'
@@ -296,8 +300,8 @@ while [ $# != 0 ]; do
 				if type cp > /dev/null 2>&1 && type awk > /dev/null 2>&1 && type sed > /dev/null 2>&1 && type diff > /dev/null 2>&1; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
 				#gtkdialog
 				TMP=`$GTKDIALOG --version 2>> $STDERR | grep -F version | cut -d' ' -f3`
-				echo -n '    gtkdialog                       |http://www.murga-linux.com/puppy/viewtopic.php?t=38608|| >= 0.8.4' >> $WORKDIR/dependencies
-				if type gtkdialog > /dev/null 2>&1 && [[ $((10#${TMP//./})) -ge 84 ]]; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
+				echo -n "    $GTKDIALOG                       |http://www.murga-linux.com/puppy/viewtopic.php?t=38608|| >= 0.8.4" >> $WORKDIR/dependencies
+				if type $GTKDIALOG > /dev/null 2>&1 && [[ $((10#${TMP//./})) -ge 84 ]]; then echo -e "|OK|" >> $WORKDIR/dependencies; else echo -e "|MISSING|" >> $WORKDIR/dependencies; fi
 				#ffmpeg is not that straight forward since compiling from git gives another output
 				TMP=`ffmpeg -version 2>> $STDERR | grep -iF 'ffmpeg' | cut -d' ' -f3 | cut -d'-' -f1`
 				echo -n '    ffmpeg                          |https://ffmpeg.org/|| >= 1.2  ' >> $WORKDIR/dependencies
@@ -423,7 +427,7 @@ if [ -s $WORKDIR/MY_PID ]; then
 fi
 echo $$ > $WORKDIR/MY_PID #store for next instance (new PMUSIC_FRONTEND or new startup without quit)
 PS="$(LC_ALL=C ps -eo pid,s,command)"
-for I in `LC_ALL=C grep -i pmusic <<< "$PS" | grep -F gtkdialog | grep -vF radiohits | awk '{print $1}'`; do kill -9 $I 2>> $STDERR; done
+for I in `LC_ALL=C grep -i pmusic <<< "$PS" | grep -F $GTKDIALOG | grep -vF radiohits | awk '{print $1}'`; do kill -9 $I 2>> $STDERR; done
 
 if [ "$NOGUI" != "true" ]; then
 	#check for data storage

```

~The very last change is the only one that's concerning because gtkdialog may refer to itself internally as 'gtkdialog', no matter the name of the exec.~

EDIT: Yes that strike thru was a problem, revised patch posted.

01micko commented:
Ok, patch works - built from woof, tested in qemu - great sound!

![qemu-pmusic(1)](https://user-images.githubusercontent.com/1019119/117559186-cb63ae80-b072-11eb-8fb6-3b689ad105f6.png)

dimkr commented:
~~Where is gtk2dialog coming from? Maybe we should change the gtkdialog petbuild so it builds against both GTK+ 2 and 3.~~

#2254

dimkr committed:
add a GTK+ 2 build of gtkdialog alongside the GTK+ 3 build

zigbert commented:
@01micko Of course add your patch. 
I first misunderstood, and thought you wanted me to add the changes (for some reason...) - Therefor my initial commit this morning.

01micko commented:
@zigbert no problem - I see you added part of it - I'll complete the job :+1:

Commit:
fix some quicksetup splash screens re #2248

Commit:
Merge pull request #2254 from dimkr/feature/gtk2dialog

Add a GTK+ 2 build of gtkdialog alongside the GTK+ 3 build

Commit:
Merge branch 'testing' into gtk3

Commit:
Merge branch 'testing' into gtk3: update slacko64

Commit:
fix typo

Commit:
bump slacko64 version

Commit:
change theme for slacko64

This shows off stardust @zigbert :-)

01micko commented:
[slacko64-8.2](https://forum.puppylinux.com/viewtopic.php?p=24908#p24908) built on bare metal from the latest `gtk3` branch available for dowmload.

dimkr commented on woof-code/rootfs-packages/ptheme/usr/sbin/ptheme:
```diff
@@ -46,7 +46,7 @@ build_menus (){
 	MENUITEMS_JWM_TRAY="$(build_menu /usr/share/jwm/tray_templates $WORKDIR/ptheme-jwm_tray PTHEME_JWM_TRAY)"
 	MENUITEMS_JWM_BUTTONS="$(build_menu /usr/share/jwm/themes_window_buttons/ $WORKDIR/ptheme-jwm_buttons PTHEME_JWM_BUTTONS)"
-	MENUITEMS_MOUSE="$(find -L $HOME/.icons -mindepth 1 -maxdepth 1 -type d -printf "<menuitem label=\"%f\"><action>echo %f > $WORKDIR/ptheme-mouse</action><action>refresh:PTHEME_MOUSE</action></menuitem>" | sort | grep -vE "ROX|default" &)"
+	MENUITEMS_MOUSE="$(find $HOME/.icons -mindepth 1 -maxdepth 1 -type d -printf "<menuitem label=\"%f\"><action>echo %f > $WORKDIR/ptheme-mouse</action><action>refresh:PTHEME_MOUSE</action></menuitem>" | sort | grep -vE "ROX|default" &)"
```
Why is this -L removed?

dimkr replied:
https://github.com/puppylinux-woof-CE/woof-CE/commit/0e9de2dcbac3daf4aa53164aa782c0a72f7d1d2f

01micko replied:
PEBKAC (here) - fixed now
58435b3

Commit:
Merge branch 'testing' into gtk3

Commit:
fix bad deletion in ptheme

Commit:
update startdust bright gtk-3.0 themes

- fixes the bug where 'relief="2"' didn't work
- fixes the scrollbar icons at top/bottom left/right
re #2248

Commit:
improve the look of ptheme GUI

refer to f6385c7 for theme changes and #2248

01micko commented:
We're getting closer  -

![gtk3-apps](https://user-images.githubusercontent.com/1019119/118344581-74505480-b51e-11eb-8486-10e139fd0798.png)

dimkr commented:
> We're getting closer -

Nice!

I'm seeing some issues with Flat-grey-rounded now - some widgets are too small with straight corners now, and stuff like MATE panel doesn't work well (window list item color doesn't change on hover).

01micko commented:
> > We're getting closer -
> 
> Nice!
> 
> I'm seeing some issues with Flat-grey-rounded now - some widgets are too small with straight corners now, and stuff like MATE panel doesn't work well (window list item color doesn't change on hover).

There are  _lots_  of issues! I haven't committed anything yet because there are tiny issues in lots of different apps. I'm still learning how to control CSS with gtk. It's similar but definitely not as forgiving as web development.

01micko commented:
BTW @dimkr what distro are you running mate on?

dimkr commented:
> BTW @dimkr what distro are you running mate on?

Ubuntu MATE. But it doesn't matter - even stuff like the "open file" dialog is messed and it shouldn't be an Ubuntu-specific problem. You can see that the padding is gone, and the square widgets look out of place.

(Maybe, every "flat" widget is messed up because the CSS selector has changed)

![flat](https://user-images.githubusercontent.com/1471149/118352240-067d4a80-b569-11eb-90c3-cf31c3860429.png)

01micko commented:
I'll get something going with MATE and take a look.

Commit:
Merge branch 'testing' into gtk3

Commit:
fix up and clean up Startdust bright for gtk3 #2248

Commit:
fix Stardust touch gtk3 #2248

Commit:
Add Startdust dark mouse gtk3 theme re #2248

Commit:
⇧Add stardust dark touch gtk3 theme re #2248

Commit:
Update ptheme for gtk3
- improved UI
- don't produce gtk_rc2 files
- send errors to /dev/null

Commit:
fix jwm bookmark tool for gtk3

(and try not to break it for gtk2)

Commit:
Attempt to fix menumanager for gtk3 re #2248
( again trying not to break gtk2 o_O )

Commit:
Add info to quicksetup

This is in the 'i' info button informing the user that they can use keyboard shortcuts to speed up navigation of the large combobox lists.

dimkr commented:
@01micko Why not merge it now?

01micko commented:
Be my guest @dimkr
